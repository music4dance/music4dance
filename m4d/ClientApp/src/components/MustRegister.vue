<template>
  <div v-if="!menuContext.isAuthenticated">
    {{ title }}
    <ul>
      <li>
        <a :href="menuContext.getAccountLink('register')">Click here to register.</a>
      </li>
      <li>
        If you're already registered,
        <a :href="menuContext.getAccountLink('login')">click here to log in.</a>
      </li>
    </ul>
  </div>
</template>

<script setup lang="ts">
import type { MenuContext } from "@/models/MenuContext";

defineProps<{ title: string; menuContext: MenuContext }>();
</script>
