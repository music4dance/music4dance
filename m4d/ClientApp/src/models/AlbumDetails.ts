import { jsonMember, jsonObject } from "typedjson";
import { PurchaseEncoded } from "./Purchase";

@jsonObject
export class AlbumDetails {
  @jsonMember(Number) public index?: number;
  @jsonMember(String) public name?: string;
  @jsonMember(Number) public track?: number;
  @jsonMember(String) public publisher?: string;
  @jsonMember(PurchaseEncoded) public purchase!: PurchaseEncoded;

  public constructor(init?: Partial<AlbumDetails>) {
    Object.assign(this, init);
    if (!this.purchase) {
      this.purchase = new PurchaseEncoded();
    }
  }
}
