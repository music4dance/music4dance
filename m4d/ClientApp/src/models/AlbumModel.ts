import { jsonMember, jsonObject } from "typedjson";
import { SongListModel } from "./SongListModel";

@jsonObject
export class AlbumModel extends SongListModel {
  @jsonMember(String) public title!: string;
  @jsonMember(String) public artist?: string;
}
