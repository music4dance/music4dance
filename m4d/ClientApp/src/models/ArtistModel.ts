import { jsonMember, jsonObject } from "typedjson";
import { SongListModel } from "./SongListModel";

@jsonObject
export class ArtistModel extends SongListModel {
  @jsonMember(String) public artist!: string;
}
