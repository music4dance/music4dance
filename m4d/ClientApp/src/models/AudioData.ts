import { jsonMember, jsonObject } from "typedjson";

@jsonObject
export class AudioData {
  @jsonMember(Number) public beatsPerMeasure?: number;
  @jsonMember(Number) public beatsPerMinute?: number;
  @jsonMember(Number) public danceability?: number;
  @jsonMember(Number) public energy?: number;
  @jsonMember(Number) public valence?: number;
}
