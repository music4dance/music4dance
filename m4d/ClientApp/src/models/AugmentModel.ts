import { jsonMember, jsonObject } from "typedjson";

@jsonObject
export class AugmentModel {
  @jsonMember(String) public title?: string;
  @jsonMember(String) public artist?: string;
  @jsonMember(String) public id?: string;
}
