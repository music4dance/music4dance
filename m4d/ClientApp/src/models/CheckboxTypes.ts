import { wordsToKebab } from "@/helpers/StringHelpers";
import type { CheckboxOption, CheckboxValue } from "bootstrap-vue-next";

// INTO-TODO: See if we can merge ListOption into these

export interface Comparable {
  equals(that?: unknown): boolean;
}

export function isComparable(value: Comparable | unknown): value is Comparable {
  return (value as Comparable).equals !== undefined;
}

export function textFromOptions(options: CheckboxOption[]): string[] {
  return options.map(({ text }) => text);
}

export function valuesFromOptions(options: CheckboxOption[]): CheckboxValue[] {
  return options.map(({ value }) => value);
}

export function textFromValues(values: CheckboxValue[], all: CheckboxOption[]): string[] {
  return all.filter((o) => values.includes(o.value)).map((o) => o.text);
}

export function optionsFromText(text: string[]): CheckboxOption[] {
  return text.map((t) => ({ text: t, value: wordsToKebab(t) }));
}
