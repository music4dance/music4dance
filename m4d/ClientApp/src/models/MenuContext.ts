import type { AxiosInstance } from "axios";
import axios from "axios";

export interface MenuContextInterface {
  helpLink?: string;
  userName?: string;
  userId?: string;
  roles?: string[];
  indexId?: string;
  expiration?: Date;
  updateMessage?: string;
  marketingMessage?: string;
  xsrfToken?: string;
}

export class MenuContext implements MenuContextInterface {
  public helpLink?: string;
  public userName?: string;
  public userId?: string;
  public roles?: string[];
  public indexId?: string;
  public expiration?: Date;
  public updateMessage?: string;
  public marketingMessage?: string;
  public xsrfToken?: string;

  public constructor(init?: MenuContextInterface) {
    Object.assign(this, init);
  }

  public get isAdmin(): boolean {
    return !!this.roles?.find((r) => r === "dbAdmin");
  }

  public get isPremium(): boolean {
    return !!this.roles?.find((r) => r === "premium" || r === "trial");
  }

  public get canTag(): boolean {
    return !!this.roles?.find((r) => r === "canTag");
  }

  public get canEdit(): boolean {
    return !!this.roles?.find((r) => r === "canEdit");
  }

  public get isBeta(): boolean {
    return !!this.roles?.find((r) => r === "beta");
  }

  public get isAuthenticated(): boolean {
    return !!this.userName;
  }

  public get daysToExpiration(): number | undefined {
    if (!this.expiration) {
      return undefined;
    }
    const timeDiff = this.expiration.getTime() - Date.now();
    const dayDiff = timeDiff / (1000 * 60 * 60 * 24);
    return dayDiff > 0 ? dayDiff : 0;
  }

  public hasRole(role: string): boolean {
    return !!this.roles?.includes(role);
  }

  public getAccountLink(page: string): string {
    return `/identity/account/${page}?returnUrl=${window.location.pathname}?${window.location.search}`;
  }

  public get axiosXsrf(): AxiosInstance {
    return axios.create({
      headers: { RequestVerificationToken: this.xsrfToken },
    });
  }
}
