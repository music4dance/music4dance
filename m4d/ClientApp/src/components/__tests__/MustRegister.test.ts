import { mount } from "@vue/test-utils";
import { describe, expect, test } from "vitest";
import { MenuContext } from "@/models/MenuContext";

import MustRegister from "../MustRegister.vue";

describe("MustRegister.vue", () => {
  test("Checks the links", () => {
    const wrapper = mount(MustRegister, {
      props: { title: "Test", menuContext: new MenuContext() },
    });
    expect(wrapper.findAll("a").length).toBe(2);
  });

  test("Renders the MustRegister component", () => {
    const wrapper = mount(MustRegister, {
      props: { title: "Test", menuContext: new MenuContext() },
    });
    expect(wrapper.html()).toMatchSnapshot();
  });
});
