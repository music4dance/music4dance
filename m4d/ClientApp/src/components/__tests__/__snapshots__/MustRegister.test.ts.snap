// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`MustRegister.vue > Renders the MustRegister component 1`] = `
"<div>Test <ul>
    <li><a href=\\"/identity/account/register?returnUrl=/?\\">Click here to register.</a></li>
    <li> If you're already registered, <a href=\\"/identity/account/login?returnUrl=/?\\">click here to log in.</a></li>
  </ul>
</div>"
`;
